### ISSUE TITLE
What's this HTTPS business exactly?

### ISSUE THREAD

**Issue by egyptiankarim** [ericahhs/ericahhs.github.io#3]
@ericahhs, some things to think about:

- What's the difference between `HTTP` and `HTTPS`, exactly?
- What's a "certificate" and how is it related to how `HTTPS` works?
- Does our GitHub Page have an `HTTPS` endpoint? If not, is it possible to set one up?

Let's have a conversation about these things here, and let's tag in @DDELS to follow along.

**Comment by ericahhs** [comment]
Hey @egyptiankarim!

- HTTP stands for Hypertext Transfer Protocol.  When the 's' extension is added, it means secure.  When the extension is added, a layer of security (encryption) is placed on the data being transferred through a secure socket layer (SSL).

- A certificate is a form of verification/identification. It binds a key that activates the protocol and allows a secure connection when on a browser.

- The GitHub Page has an HTTPS endpoint currently, but I noticed it is possible to disable it in the 'Settings' tab.

**Comment by egyptiankarim** [comment]
> HTTP stands for Hypertext Transfer Protocol. When the 's' extension is added, it means secure. When the extension is added, a layer of security (encryption) is placed on the data being transferred through a secure socket layer (SSL).

Correct! Modern practices for negotiating secure HTTP(S) connections between clients and servers are moving away from SSL and towards TLS, but the principles are all still the same. A client calls a server and says "I want to connect to you securely", the server replies back with the appropriate certificates and a key exchange is made, and then traffic is encrypted according to the details of that transaction.

> A certificate is a form of verification/identification. It binds a key that activates the protocol and allows a secure connection when on a browser.

You've got it! How are certificates created/issued? How is the validity of a certificate established?

> The GitHub Page has an HTTPS endpoint currently, but I noticed it is possible to disable it in the 'Settings' tab.

Correct again! We also have the option to "enforce" HTTPS connections. Do a little a digging and see if you can find out what's implied by such enforcement.

**Comment by ericahhs** [comment]
> How are certificates created/issued? How is the validity of a certificate established?

Certificates are issued by certification authorities (CAs) who validate the identity of the requester before and while the certificate is in use.

The validity of a certificate is established by using a technique known as "domain validation."  Domain validation techniques are used to prove that the certificate applicant controls a given domain name, not any information about the identity of the applicant.

> We also have the option to "enforce" HTTPS connections. Do a little a digging and see if you can find out what's implied by such enforcement.

In Settings, it states that when HTTPS is enforced, the site will only be served over HTTPS.

**Comment by egyptiankarim** [comment]
> Certificates are issued by certification authorities (CAs) who validate the identity of the requester before and while the certificate is in use.

Interesting! See if you can find examples of how this process has gone awry in the past (there's a lot of pretty interesting stories, actually). Also, Google around for information about [Let's Encrypt](https://letsencrypt.org), and see if you can find out how they're trying to resolve some of the shortcomings of traditional CAs.

> In Settings, it states that when HTTPS is enforced, the site will only be served over HTTPS.

For sure. Here's a hint: Google around for [HSTS](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security), and see if you can sort out how it works. Bonus points if you can identify a quick and easy way to test and see if a web server has implemented it. Bonus bonus points if you read up on the role all of this plays in modern federal government initiatives (here's another hint: [The HTTPS-Only Standard](https://https.cio.gov)).

**Comment by ericahhs** [comment]
Some issues with the traditional CA process include: Loss of cryptographic keys, many fraudulent certificates, and attackers being able to compromise keys or certificates for a long time (because some certificates have validity for 1-3 years).

> Google around for information about Let's Encrypt, and see if you can find out how they're trying to resolve some of the shortcomings of traditional CAs

Let's Encrypt is trying to fix this issue by using automation for many steps within the process to limit the possibility of fraudulent acts.  They let users and organizations obtain, renew, and manage SSL/TLS certificates; and their certificates can be used by websites to enable secure HTTPS connections.

> Google around for HSTS, and see if you can sort out how it works

HTTP Strict Transport Security (HSTS) is a policy mechanism that helps protect websites by allowing the web servers to interact **only** with a secure HTTPS connection.

> Bonus points if you can identify a quick and easy way to test and see if a web server has implemented it

If HSTS is enabled, web browsers automatically change any insecure requests (http://) to secure requests (https://). You can change the https in the URL to http and if it has HSTS then it will automatically reload as https.

> Bonus bonus points if you read up on the role all of this plays in modern federal government initiatives

All publicly accessible Federal websites and web services are required to only provide service through a secure connection (HTTPS).  There are many guidelines that federal agencies must be in compliance with.

**Comment by egyptiankarim** [comment]
@ericahhs I think you've got it!

> Let's Encrypt is trying to fix this issue by using automation for many steps within the process to limit the possibility of fraudulent acts. They let users and organizations obtain, renew, and manage SSL/TLS certificates; and their certificates can be used by websites to enable secure HTTPS connections.

Automation is a huge deal. The short lifespans of Let's Encrypt certificates play a big role too, and target longstanding problems with certificate revocation. Also, it's kind of great that the Let's Encrypt folks approached the problem of getting the internet encrypted with a mind towards low cost.

> All publicly accessible Federal websites and web services are required to only provide service through a secure connection (HTTPS). There are many guidelines that federal agencies must be in compliance with.

Lots of information out there, for sure. But one primarily important document is [OMB M-15-13, Policy to Require Secure Connections across Federal Websites and Web Services](https://www.whitehouse.gov/sites/whitehouse.gov/files/omb/memoranda/2015/m-15-13.pdf). Give it a once over and think about how the Federal Government formulates technical policy and initiates it into action; it's kind of neat!

Okay. Last steps:

- Take a look at [An Introduction to HTTPS](https://www.youtube.com/watch?v=d2GmcPYWm5k), [Implementing HTTPS](https://www.youtube.com/watch?v=rnM2qAfEG-M), [Migrating .gov to HTTPS](https://www.youtube.com/watch?v=X5H8JRULDOo). [Eric Mill](https://github.com/konklone) and [Gray Brooks](https://github.com/gbinal) are supremely savvy and have been influential in how the Federal Government securely manages web services.
- Surf over to [cisagov](https://github.com/cisagov)/[pshtt](https://github.com/cisagov/pshtt). The Cyber and Infrastructure Security Agency is behind multiple open source tools in use across the Federal Government, and I think it'll be great for you to work on some contributions during your summer internship (and beyond!). I might suggest starting off with a fork of `pshtt`, browsing some of the issues and pull requests, and seeing if you can get the code to successfully run locally on your workstation (we can sync up about all of this during our next group meeting).
